src/main.rs

```diff
@@ -71,203 +84,227 @@ impl Player {
+            match contents[i] {
+                Normal(c) => r.contents = c,
+                Wumpus => {
+                    r.contents = Empty;
+                    r.wumpus = true;
+                }
+                Player => {
+                    r.contents = Empty;
+                    player_loc = Some(i);
+                }
+            }
         }
 
-        maze
+        (Maze { rooms, rng }, player_loc.unwrap())
     }
 
-    fn rnd_empty_room(&mut self) -> usize {
-        let empty_rooms: Vec<_> = self.rooms.iter()
-            .filter(|n| n.dangers.is_empty())
-            .collect();
+    /// Description strings for adjacent dangers.
+    const DESCS: &'static [(RoomContents, &'static str)] = &[
+        (Pit, "You feel a cold wind blowing from a nearby cavern."),
+        (Bat, "You hear a rustling."),
+    ];
 
-        empty_rooms
-            .choose(RefCell::borrow_mut(&self.rng).deref_mut())
-            .unwrap()
-            .id
-    }
+    /// Current room description string.
+    fn describe_room(&self, room: usize) -> String {
+        let mut desc_lines = Vec::new();
+        desc_lines.push(format!("You are in room #{}", room));
 
-    fn rnd_empty_neighbour(&mut self, room: usize) -> Option<usize> {
-        let neighbour_ids = self.rooms[room].neighbour_ids();
+        for (c, d) in Maze::DESCS {
+            if self.is_danger_nearby(room, *c) {
+                desc_lines.push(d.to_string());
+            }
+        }
+
+        if self.is_wumpus_nearby(room).is_some() {
+            desc_lines.push(
+                "You smell something terrible nearby.".to_string(),
+            );
+        }
 
-        let empty_neighbours: Vec<_> = neighbour_ids.iter()
-            .filter(|&n| self.rooms[*n].dangers.is_empty())
+        let neighbors: Vec<String> = self.rooms[room]
+            .neighbors
+            .iter()
+            .map(|i| i.to_string())
             .collect();
 
-        if empty_neighbours.is_empty() {
-            return None;
-        }
+        desc_lines
+            .push(format!("Exits go to: {}", neighbors.join(", "),));
 
-        let empty_neighbour = empty_neighbours
-            .choose(RefCell::borrow_mut(&self.rng).deref_mut())
-            .unwrap();
+        desc_lines.join("\n")
+    }
 
-        Some(**empty_neighbour)
+    /// Adjacent room contains a non-wumpus danger.
+    fn is_danger_nearby(
+        &self,
+        room: usize,
+        danger: RoomContents,
+    ) -> bool {
+        self.rooms[room]
+            .neighbors
+            .iter()
+            .any(|&n| self.rooms[n].contents == danger)
     }
 
-    fn describe_room(&self, room: usize) -> String {
-        let mut description = format!("You are in room #{}", room);
+    /// Specific adjacent room contains the Wumpus.
+    fn is_wumpus_nearby(&self, room: usize) -> Option<usize> {
+        self.rooms[room]
+            .neighbors
+            .iter()
+            .cloned()
```

Comment:
What does `cloned` do here?

Reply:
Dereferences the numbers. The clone is a noop here because `usize` is copy, but it gets the types right.

Alternative is `.map(|&r| r)` but `.cloned()` is more idiomatic and may be more efficient…

It's frustrating that `.into_iter()` on a `Copy` type still adds a reference, but there you are.